components/oauth/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/token/handlers/grant/jwt/JWTBearerGrantHandler.java

```diff
@@ -0,0 +1,653 @@
+import org.wso2.carbon.identity.application.common.util.IdentityApplicationManagementUtil;
+import org.wso2.carbon.identity.oauth.cache.JWTCache;
+import org.wso2.carbon.identity.oauth.cache.JWTCacheEntry;
+import org.wso2.carbon.identity.oauth.config.OAuthServerConfiguration;
+import org.wso2.carbon.identity.oauth2.IdentityOAuth2Exception;
+import org.wso2.carbon.identity.oauth2.model.RequestParameter;
+import org.wso2.carbon.identity.oauth2.token.OAuthTokenReqMessageContext;
+import org.wso2.carbon.identity.oauth2.token.handlers.grant.AbstractAuthorizationGrantHandler;
+import org.wso2.carbon.idp.mgt.IdentityProviderManager;
+import org.wso2.carbon.utils.multitenancy.MultitenantConstants;
+
+import java.security.cert.CertificateException;
+import java.security.cert.X509Certificate;
+import java.security.interfaces.RSAPublicKey;
+import java.text.ParseException;
+import java.util.Date;
+import java.util.List;
+import java.util.Map;
+
+/**
+ *  Class to handle JSON Web Token(JWT) grant type
+ */
+public class JWTBearerGrantHandler extends AbstractAuthorizationGrantHandler {
+
+    private static Log log = LogFactory.getLog(JWTBearerGrantHandler.class);
+    private JWTCache jwtCache;
+
+    // JWT Profile for Oauth2 constants
+    private static final String OAUTH_JWT_BEARER_GRANT_TYPE = "urn:ietf:params:oauth:grant-type:jwt-bearer";
+    private static final String OAUTH_JWT_BEARER_GRANT_ERROR = "invalid_grant";
+    private static final String OAUTH_JWT_BEARER_ISSUER = "issuer";
+    private static final String OAUTH_JWT_ASSERTION = "assertion";
+
+    // allowed number of minutes a token can be old for
+    private static final int REJECT_TOKEN_ISSUED_BEFORE = 30;
+
+    // whether a list of used JWT IDs are cached/persisted
+    private static final boolean CACHE_USED_JTI = true;
+    private static String tenantDomain;
+
+    // initialize the JWT cache
+    public void init() throws IdentityOAuth2Exception {
+        super.init();
+        if (CACHE_USED_JTI) {
+            this.jwtCache = JWTCache.getInstance();
+        }
+    }
+
+
+    /**
+     * We're validating the JWT token that we receive from the request. Through the assertion parameter is the POST
+     * request. A request format that we handle here looks like,
+     * <p/>
+     * POST /token.oauth2 HTTP/1.1
+     * Host: as.example.com
+     * Content-Type: application/x-www-form-urlencoded
+     * <p/>
+     * grant_type=urn%3Aietf%3Aparams%3Aoauth%3Agrant-type%3Ajwt-bearer
+     * &assertion=eyJhbGciOiJFUzI1NiJ9.
+     * eyJpc3Mi[...omitted for brevity...].
+     *
+     * @param tokReqMsgCtx Token message request context
+     * @return true if validation is successful, false otherwise
+     * @throws org.wso2.carbon.identity.oauth2.IdentityOAuth2Exception
+     */
+    @Override
+    public boolean validateGrant(OAuthTokenReqMessageContext tokReqMsgCtx) throws IdentityOAuth2Exception {
+
+        super.validateGrant(tokReqMsgCtx);
+
```

Comment:
Validate the presence of all the required attributes first and then proceed. Provide a meaningful error message.